pkg/blockbuilder/scheduler/scheduler.go

```diff
@@ -71,47 +77,51 @@ func (s *BlockBuilderScheduler) starting(ctx context.Context) error {
 	}
 
 	s.adminClient = kadm.NewClient(kc)
-
-	// The startup process for block-builder-scheduler entails learning the state of the world:
-	//  1. obtain an initial set of offset info from Kafka
-	//  2. listen to worker updates for a while to learn what the previous scheduler knew
-	// When both of those are complete, we transition from observation mode to normal operation.
-
-	var wg sync.WaitGroup
-	wg.Add(2)
-
-	go func() {
-		defer wg.Done()
-		lag, err := s.fetchLag(ctx)
-		if err != nil {
-			panic(err)
-		}
-		s.committed = commitOffsetsFromLag(lag)
-	}()
-	go func() {
-		defer wg.Done()
-		select {
-		case <-time.After(s.cfg.StartupObserveTime):
-		case <-ctx.Done():
-		}
-	}()
-
-	wg.Wait()
-
-	if err := ctx.Err(); err != nil {
-		return err
-	}
-
-	s.completeObservationMode()
 	return nil
 }
 
 func (s *BlockBuilderScheduler) stopping(_ error) error {
+	if err := s.flushOffsetsToKafka(context.Background()); err != nil {
+		level.Error(s.logger).Log("msg", "failed to flush offsets at shutdown", "err", err)
+	}
+
 	s.adminClient.Close()
 	return nil
 }
 
 func (s *BlockBuilderScheduler) running(ctx context.Context) error {
+	level.Info(s.logger).Log("msg", "entering observation mode")
+
+	observeComplete := time.After(s.cfg.StartupObserveTime)
+
+	c, err := s.fetchCommittedOffsets(ctx)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "failed to fetch committed offsets", "err", err)
+		s.metrics.fetchOffsetsFailed.Inc()
+		return fmt.Errorf("fetch committed offsets: %w", err)
+	}
+	level.Debug(s.logger).Log("msg", "loaded initial committed offsets", "offsets", offsetsStr(c))
+	s.mu.Lock()
+	s.committed = c
+	s.mu.Unlock()
+
+	// Wait for StartupObserveTime to pass.
+	select {
+	case <-observeComplete:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	// Now we can transition to normal operation.
+
+	s.completeObservationMode(ctx)
```

Comment:
(another nit) This chunk before was a part of the "starting" step of the service's state machine? Why do you think it should go to the "running" step?

Reply:
it needs to receive RPCs through the Service during this time, but it won’t be added as a service endpoint until it is healthy. (which is when it’s in the Running state.)